**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №8

Набор данных: credit_risk.csv

Атрибуты: Id, Age, Income, Home, Emp_length, Intent, Amount, Rate, Status, Percent_income, Default, Cred_length

### 1.Чтение файла (набора данных)

In [20]:
import pandas as pd
from IPython.display import display
df = pd.read_csv("C:\\Users\\vanya\\Downloads\\credit_risk.csv")
print("File is read")

File is read


### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [2]:
df.head(20)

Id   Age  Income      Home  Emp_length             Intent  Amount   Rate  \
0    0  22.0   59000      RENT       123.0           PERSONAL   35000  16.02   
1    1  21.0    9600       OWN         5.0          EDUCATION    1000  11.14   
2    2  25.0    9600  MORTGAGE         1.0            MEDICAL    5500  12.87   
3    3  23.0   65500      RENT         4.0            MEDICAL   35000  15.23   
4    4  24.0   54400      RENT         8.0            MEDICAL   35000  14.27   
5    5  21.0    9900       OWN         2.0            VENTURE    2500   7.14   
6    6  26.0   77100      RENT         8.0          EDUCATION   35000  12.42   
7    7  24.0   78956      RENT         5.0            MEDICAL   35000  11.11   
8    8  24.0   83000      RENT         8.0           PERSONAL   35000   8.90   
9    9  21.0   10000       OWN         6.0            VENTURE    1600  14.74   
10  10  22.0   85000      RENT         6.0            VENTURE   35000  10.37   
11  11  21.0   10000       OWN         2.0    HOMEIMPROVEMENT    4500   8.63   
12  12  23.0   95000      RENT         2.0            VENTURE   35000   7.90   
13  13  26.0  108160      RENT         4.0          EDUCATION   35000  18.39   
14  14  23.0  115000      RENT         2.0          EDUCATION   35000   7.90   
15  15  23.0  500000  MORTGAGE         7.0  DEBTCONSOLIDATION   30000  10.65   
16  16  23.0  120000      RENT         0.0          EDUCATION   35000   7.90   
17  17  23.0   92111      RENT         7.0            MEDICAL   35000  20.25   
18  18  23.0  113000      RENT         8.0  DEBTCONSOLIDATION   35000  18.25   
19  19  24.0   10800  MORTGAGE         8.0          EDUCATION    1750  10.99   

    Status  Percent_income Default  Cred_length  
0        1            0.59       Y            3  
1        0            0.10       N            2  
2        1            0.57       N            3  
3        1            0.53       N            2  
4        1            0.55       Y            4  
5        1            0.25       N            2  
6        1            0.45       N            3  
7        1            0.44       N            4  
8        1            0.42       N            2  
9        1            0.16       N            3  
10       1            0.41       N            4  
11       1            0.45       N            2  
12       1            0.37       N            2  
13       1            0.32       N            4  
14       0            0.30       N            4  
15       0            0.06       N            3  
16       0            0.29       N            4  
17       1            0.32       N            4  
18       1            0.31       N            4  
19       1            0.16       N            2

2.2 Оценка данных с помощью метода info.

In [3]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              652 non-null    int64  
 1   Age             652 non-null    float64
 2   Income          652 non-null    int64  
 3   Home            652 non-null    object 
 4   Emp_length      643 non-null    float64
 5   Intent          652 non-null    object 
 6   Amount          652 non-null    int64  
 7   Rate            586 non-null    float64
 8   Status          652 non-null    int64  
 9   Percent_income  652 non-null    float64
 10  Default         652 non-null    object 
 11  Cred_length     652 non-null    int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 61.2+ KB


2.3 Оценка данных с помощью метода describe.

In [4]:
# оцените числовые столбцы с помощью describe
df.describe(include='all')

Id         Age         Income  Home  Emp_length     Intent  \
count   652.000000  652.000000     652.000000   652  643.000000        652   
unique         NaN         NaN            NaN     5         NaN          6   
top            NaN         NaN            NaN  RENT         NaN  EDUCATION   
freq           NaN         NaN            NaN   415         NaN        145   
mean    325.495399   24.285276   90008.855828   NaN    4.601866        NaN   
std     188.352535    7.841061   69530.312525   NaN    7.281616        NaN   
min       0.000000   21.000000    9600.000000   NaN    0.000000        NaN   
25%     162.750000   23.000000   44000.000000   NaN    2.000000        NaN   
50%     325.500000   24.000000   69998.000000   NaN    4.000000        NaN   
75%     488.250000   25.000000  128499.000000   NaN    7.000000        NaN   
max     649.000000  144.000000  500000.000000   NaN  123.000000        NaN   

              Amount        Rate      Status  Percent_income Default  \
count     652.000000  586.000000  652.000000      652.000000     652   
unique           NaN         NaN         NaN             NaN       3   
top              NaN         NaN         NaN             NaN       N   
freq             NaN         NaN         NaN             NaN     499   
mean    18800.996933   12.289471    0.601227        0.279831     NaN   
std      9152.896250    3.272772    0.490022        0.150231     NaN   
min      1000.000000    5.420000    0.000000        0.010000     NaN   
25%     10000.000000   10.250000    0.000000        0.160000     NaN   
50%     21850.000000   12.180000    1.000000        0.280000     NaN   
75%     25000.000000   14.720000    1.000000        0.380000     NaN   
max     35000.000000   21.210000    1.000000        0.830000     NaN   

        Cred_length  
count    652.000000  
unique          NaN  
top             NaN  
freq            NaN  
mean       3.012270  
std        0.808843  
min        2.000000  
25%        2.000000  
50%        3.000000  
75%        4.000000  
max        4.000000


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**


 ---


 2.4 Оценка названий столбцов

In [5]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns.to_list()

['Id',
 'Age',
 'Income',
 'Home',
 'Emp_length',
 'Intent',
 'Amount',
 'Rate',
 'Status',
 'Percent_income',
 'Default',
 'Cred_length']

In [30]:
# переименование при необходимости
df = df.rename(columns={'Home': 'Home_status'})
display(df.columns.to_list())

['Id',
 'Age',
 'Income',
 'Home_status',
 'Emp_length',
 'Intent',
 'Amount',
 'Rate',
 'Status',
 'Percent_income',
 'Default',
 'Cred_length']

### 3. Проверка пропусков

In [21]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
res_dict = {}
for i, value in df.isna().sum().items():
    if value != 0:
        res_dict[i] = None
for i in res_dict:
    if df[i].dtypes == 'float64':
        res_dict[i] = df[i].median()
    elif df[i].dtypes == 'int64':
        res_dict[i] = int(df[i].median())
    elif df[i].dtypes == 'object':
        res_dict[i] = df[i].moda(0)
df.fillna(res_dict, inplace=True)

display(df.isna().sum())
        

Id                0
Age               0
Income            0
Home              0
Emp_length        0
Intent            0
Amount            0
Rate              0
Status            0
Percent_income    0
Default           0
Cred_length       0
dtype: int64


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [23]:
print(df.duplicated().sum())
for i, value in df.duplicated().items():
    print(i, value) if value else None

2
650 True
651 True


In [9]:
# удалите дубликаты, если они есть
df.drop_duplicates(inplace=True)
print(df.duplicated().sum())

0


#### Проверка неявных дубликатов

In [10]:
colomns = ['Home_status', 'Intent', 'Default']
for colomn in colomns:
    print(colomn)
    for i in df[colomn].unique():
        print(f"\t{i}")


Home_status
	RENT
	OWN
	MORTGAGE
	RENET
	OTHER
Intent
	PERSONAL
	EDUCATION
	MEDICAL
	VENTURE
	HOMEIMPROVEMENT
	DEBTCONSOLIDATION
Default
	Y
	N
	No


In [11]:
# удалите дубликаты, если они есть
df.replace({'Home_status': {'RENET': 'RENT'}}, inplace=True)
df.replace({'Default': {'No': 'N'}}, inplace=True)

colomns = ['Home_status', 'Intent', 'Default']
for colomn in colomns:
    print(colomn)
    for i in df[colomn].unique():
        print(f"\t{i}")

Home_status
	RENT
	OWN
	MORTGAGE
	OTHER
Intent
	PERSONAL
	EDUCATION
	MEDICAL
	VENTURE
	HOMEIMPROVEMENT
	DEBTCONSOLIDATION
Default
	Y
	N


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [24]:
display(df.head(2))
display(df.dtypes)

Id   Age  Income  Home  Emp_length     Intent  Amount   Rate  Status  \
0   0  22.0   59000  RENT       123.0   PERSONAL   35000  16.02       1   
1   1  21.0    9600   OWN         5.0  EDUCATION    1000  11.14       0   

   Percent_income Default  Cred_length  
0            0.59       Y            3  
1            0.10       N            2

Id                  int64
Age               float64
Income              int64
Home               object
Emp_length        float64
Intent             object
Amount              int64
Rate              float64
Status              int64
Percent_income    float64
Default            object
Cred_length         int64
dtype: object

In [27]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности
res_dict = {'Id': 'int64', 'Age': 'int64', 'Income': 'int64', 'Home_status': 'object', 
            'Emp_length': 'int64', 'Intent': 'object', 'Amount': 'int64', 'Rate': 'float64',
            'Status': 'bool', 'Percent_income': 'float', 'Default': 'object', 'Cred_length': 'int64'}

df = df.astype(res_dict)
display(df.head(2))
display(df.dtypes)

Id  Age  Income Home_status  Emp_length     Intent  Amount   Rate  Status  \
0   0   22   59000        RENT         123   PERSONAL   35000  16.02    True   
1   1   21    9600         OWN           5  EDUCATION    1000  11.14   False   

   Percent_income Default  Cred_length  
0            0.59       Y            3  
1            0.10       N            2

Id                  int64
Age                 int64
Income              int64
Home_status        object
Emp_length          int64
Intent             object
Amount              int64
Rate              float64
Status               bool
Percent_income    float64
Default            object
Cred_length         int64
dtype: object

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

*`Формулировка задания`*
 Группировка - статус домовладения (home) по количеству дефолтов
(default)

In [14]:
# выполните группировку согласно варианту
df.groupby(['Home_status', 'Default']).size()

Home_status  Default
MORTGAGE     N          107
             Y           29
OTHER        Y            2
OWN          N           85
             Y           13
RENT         N          308
             Y          106
dtype: int64

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Можно сделать вывод о том, что наименьший процент дефолта кредита является ситуация, когда у потенциального клиента дом в сбоственности(13,3%).
На втором месте являются потенциальные клиенты с ипотекой (21,3%) и самым рисковым является вариант съемной квартиры (25,5%).

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Формулировка задания`*

Группировка - цель кредита (intent) по количеству статуса
домовладения (home). Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по убыванию столбца “count”.

In [15]:
# выполните группировку согласно варианту
df.groupby(['Intent', 'Home_status']).size().reset_index(name='count').sort_values(by='count', ascending=False)

Intent Home_status  count
6           EDUCATION        RENT     94
2   DEBTCONSOLIDATION        RENT     72
12            MEDICAL        RENT     70
19            VENTURE        RENT     70
16           PERSONAL        RENT     69
9     HOMEIMPROVEMENT        RENT     39
3           EDUCATION    MORTGAGE     34
17            VENTURE    MORTGAGE     27
10            MEDICAL    MORTGAGE     22
15           PERSONAL         OWN     22
0   DEBTCONSOLIDATION    MORTGAGE     21
18            VENTURE         OWN     20
13           PERSONAL    MORTGAGE     19
5           EDUCATION         OWN     16
8     HOMEIMPROVEMENT         OWN     16
11            MEDICAL         OWN     15
7     HOMEIMPROVEMENT    MORTGAGE     13
1   DEBTCONSOLIDATION         OWN      9
4           EDUCATION       OTHER      1
14           PERSONAL       OTHER      1

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Образование является одной из самых главных причин для взятия кредита, Процент кредитов по этой отрасли максимален, если у человека
съемное жилье или есть ипотека и находится в топ-3, если у клиента собственное жилище.
Также стоит отметить, что консолидация всех задолженностей является очень распространенной причиной взятия кредита только у людей, которые
арендуют квартиру.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Формулировка задания`*

Сводная таблица (pivot_table) - средний доход income для каждого
возраста. Отсортировать по убыванию. Округлить до двух знаков.

In [28]:
# выполните сводную таблицу согласно варианту
income_for_age = df.pivot_table(index='Age', values='Income', aggfunc='mean').sort_values(by='Income', ascending=False).round(2)
display(income_for_age)

Income
Age           
144  225000.00
26   122793.74
25   106750.02
24    88684.03
23    85275.05
123   80004.00
22    64423.88
21    42238.58

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

Исходя из итоговых данных видно, что чем выше опыт юных клиентов (от 21 до 26 лет), тем выше его платежеспособность.
Также видно возможную допущенную ошибку при заполнении данных (возраста 144 и 123 года). Рекомендуется уточнить актуальность данных,
после чего принять решение: удалить и не учитывать данные либо написать корректные значения и заново сделать сводную таблицу.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Формулировка задания`*

Сводная таблица (pivot_table) - средняя сумма кредита amount для
каждой цели (intent) - строки и возраста - столбцы . Отсортировать по
возрастанию intent. Округлить до двух знаков

In [33]:
# выполните сводную таблицу согласно варианту
mean_amount = df.pivot_table(index='Intent', columns='Age', values='Amount').sort_values(by='Intent', ascending=True).round(2)
mean_amount['sum'] = mean_amount.sum(axis=1)
display(mean_amount)

Age                      21        22        23        24        25        26  \
Intent                                                                          
DEBTCONSOLIDATION   9500.00  18602.78  20275.00  20028.57  18603.33  21947.22   
EDUCATION          12846.15  18677.34  20288.28  21052.38  19326.04  20566.67   
HOMEIMPROVEMENT     4355.00  10216.67   4000.00  19627.68  21266.67  20038.46   
MEDICAL            10700.00  18073.21  17968.48  21119.64  18114.81  20410.00   
PERSONAL            3633.33  17735.71  19241.00  21205.00  21738.00  18588.24   
VENTURE             9407.69  15515.28  22810.94  19406.73  18360.42  18850.00   

Age                    123     144        sum  
Intent                                         
DEBTCONSOLIDATION      NaN     NaN  108956.90  
EDUCATION          20400.0  6000.0  139156.86  
HOMEIMPROVEMENT        NaN     NaN   79504.48  
MEDICAL                NaN     NaN  106386.14  
PERSONAL               NaN     NaN  102141.28  
VENTURE                NaN  4800.0  109151.06

**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

Для наглядности была добавлена колонка Суммы по строкам. Образование остается самой востребованной целью для оформление кредита,
при чем несмотря на внушительную часть "сомнительных" данных (возраст 123 и 144 года), которые составляют почти 19% от общей суммы.
Самым же недорогой целью является улучшение жилищных условий. Оно остается самым низким вплоть до 24 лет, после чего ни разу не будет находится на последнем месте. Это говорит о том, что по мере взросления человек хочет все больше и больше вкладывать в обустройства своего жилища. Этот тренд должен продолжиться и далее.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод

В ходе выполнения лабораторной работы был проведён предварительный анализ и обработка набора данных credit_risk.csv, содержащего информацию о кредитных заявках. Набор данных включал такие атрибуты, как возраст заёмщика, уровень дохода, цель кредита, статус владения жильем, процент дохода, выделяемый на выплаты, а также факт наличия дефолта.
Была выполнена следующая последовательность обработки данных:
1. Первичный обзор структуры данных с использованием методов head(), info() и describe(), что позволило оценить общий объём данных, наличие пропусков и распределение значений.
2. Корректировка названий столбцов для улучшения читаемости (например, столбец Home был переименован в Home_status).
3. Заполнение пропусков: обнаружены и заполнены пропущенные значения в столбцах Emp_length и Rate.
4. Удаление дубликатов: выявлены и удалены как явные, так и неявные дубликаты, в том числе вызванные разным написанием категориальных значений.
5. Корректировка типов данных: приведены к целочисленному типу столбцы Age и Emp_length, а также преобразован столбец Status в булевый тип.
На этапе аналитической группировки данных были выявлены содержательные закономерности:

•	Наименьший уровень дефолта наблюдается среди клиентов, владеющих жильём (13,3%), тогда как арендаторы демонстрируют наибольшие риски (25,5%).

•	Основной целью получения кредита является образование, особенно среди арендаторов и лиц с ипотекой.

•	Обнаружены аномалии в данных (возрастные значения 123 и 144 года), которые могут указывать на ошибки ввода и требуют дополнительной проверки.

•	Наибольшие средние суммы кредитов приходятся на образовательные цели, тогда как наименьшие — на улучшение жилищных условий, особенно среди молодых заёмщиков.

Проведённая работа позволила не только очистить и структурировать исходные данные, но и выявить ключевые факторы, влияющие на кредитные риски и поведение заёмщиков. Полученные результаты могут быть использованы для построения прогнозных моделей и поддержки принятия решений в области кредитного скоринга.


### Дополнительное задание

**`Подробная формулировка задания`**

In [18]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***